rs/consensus/src/consensus/validator.rs

```diff
@@ -977,6 +1053,35 @@ impl Validator {
                 }
             }
 
+            // Disqualify rank if equivocation was found. If there already
+            // exists a validated block of the same rank as the current
+            // proposal, we must generate an equivocation proof.
+            if let Some(existing_metadata) =
+                valid_qualified_ranks.get(proposal.height(), proposal.rank())
+            {
+                // Ensure the proposal has a different hash from the validated
+                // block of same rank. Then we can construct the proof.
+                if proposal.content.get_hash().get_ref() != existing_metadata.content.hash() {
+                    change_set.push(ChangeAction::AddToValidated(ValidatedArtifact {
+                        msg: ConsensusMessage::EquivocationProof(EquivocationProof {
+                            signer: proposal.signature.signer,
+                            version: proposal.content.version().clone(),
+                            height: proposal.height(),
+                            subnet_id: self.replica_config.subnet_id,
+                            hash1: proposal.content.get_hash().clone(),
+                            signature1: proposal.signature.signature.clone(),
+                            hash2: CryptoHashOf::new(existing_metadata.content.hash().clone()),
+                            signature2: existing_metadata.signature.signature,
+                        }),
+                        timestamp: self.time_source.get_relative_time(),
+                    }));
+                    valid_qualified_ranks.remove(proposal.height(), proposal.rank());
+                    disqualified_ranks.add(&proposal);
+                    // Blocks from disqualified ranks can be ignored at this point
+                    continue;
+                }
```

Comment:
Should we also handle the else case explicitly? Or is it fine to validate the block?

Reply:
So far we've always validated duplicate blocks, which seems fine to me.

Reply:
Letting duplicates be validated seems fine to me.